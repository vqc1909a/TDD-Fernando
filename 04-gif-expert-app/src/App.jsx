import { useState } from 'react'
import './App.css'
import { AddCategory, GifGridCategory } from "./components/Gif";
import { nanoid } from 'nanoid'

const DEFAULT_CATEGORIES = [
  {
    name: 'One Puch',
    id: nanoid()
  }
]
function App() {

  const [categories, setCategories] = useState(DEFAULT_CATEGORIES)

  const handleAddCategory = (category) => {
    const newCategory = {
      name: category,
      id: nanoid()
    }
    setCategories(categories => [newCategory, ...categories]);
  }

  return (
    <div>
     <h1>GifExpertApp</h1> 
     <AddCategory handleAddCategory={handleAddCategory}></AddCategory>
     {/* React recommends not use the parameter index for the keys, because it could be a problem when the list changes. The best way is use a unique key, like the id of the object brought out of the npm package or the db. */}
     <ol>
      {
        categories.map(({ id, name}) => <GifGridCategory key={id} category={name}></GifGridCategory>)
      }
     </ol>
    </div>
  )
}

export default App
