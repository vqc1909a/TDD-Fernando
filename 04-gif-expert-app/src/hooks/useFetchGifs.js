import { useEffect, useState } from "react";
import { getGifsByCategory } from "../helpers/helperGif";

export const useFetchGifs = (category) => {
  const [gifs, setGifs] = useState([]);
  const [isLoading, setIsLoading] = useState(true);
  
  //By the mode strict in the file main.jsx, the console.log show twice as many times as expected. But in production, it will be once time
  // console.log({
  //   gifs
  // })
  
  const getGifs = async () => {
    const gifs = await getGifsByCategory(category);
    setGifs(gifs);
    setIsLoading(false)
  }

  useEffect(() => {
    getGifs();
    //eslint-disable-next-line
  }, [])


  return {
    gifs,
    isLoading
  };

}
