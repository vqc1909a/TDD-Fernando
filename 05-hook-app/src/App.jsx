
function App() {

  return (
    <>
      <h1>HooksApp</h1>
    </>
  )
}

export default App
