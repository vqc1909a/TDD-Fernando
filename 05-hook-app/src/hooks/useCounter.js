import { useState } from "react"

export const useCounter = (initialValue = 10) => {
  const [counter, setCounter] = useState(initialValue); 

  const handleAddCounter = (value = 1) => {
    setCounter(counter + value);
  }

  const handleResetCounter = (value = initialValue) => {
    setCounter(value);
  }
  const handleSubtractCounter = (value = 1) => {
    if(counter === 0) return;
    setCounter(counter - value);
  }
  return {
    counter,
    handleAddCounter,
    handleResetCounter,
    handleSubtractCounter
  }
}
