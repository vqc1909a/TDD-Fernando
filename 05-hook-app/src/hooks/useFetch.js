import { useEffect, useState } from "react"

export const useFetch = (url) => {
  
  const [state, setState] = useState({
      data: [],
      isLoading: true,
      hasError: false
  })

  const getFetch = async (url) => {
    setState(state => ({ 
        ...state,
        isLoading: true,
    }))

    const resp = await fetch(url)
    const data = await resp.json()
    setState(state => ({
      ...state,       
      data,
      isLoading: false,
      hasError: false 
    }))
  }

  useEffect(() => {
    getFetch(url);
    // eslint-disable-next-line 
  }, [url])
  
  return {
    ...state
  }
}
