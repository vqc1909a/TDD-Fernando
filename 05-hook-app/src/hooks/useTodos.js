import { useCallback, useEffect, useReducer } from "react";
import { todoReducer } from "../08-useReducer/todoReducer";

const init = () => JSON.parse(localStorage.getItem('todos')) || []

export const useTodos = () => {
    //We use the third argument of the useReducer to initialize the state of the reducer    
    const [todos, dispatch] = useReducer(todoReducer, [], init); 

    //The recommended is we have my functions that changed the state in the father component
    const handleAddTodo = useCallback((todo) => {
        dispatch({
            type: "addTodo",
            payload: todo
        })
    }, []);

    const handleDeleteTodo = (id) => {
        dispatch({
            type: "deleteTodo",
            payload: id
        })
    };

    const handleToggleTodo = (id) => {
        dispatch({
            type: "toggleTodo",
            payload: id
        })
    }

    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos])

    return {
        todos,
        todosCount: todos.length,
        pendingTodosCount: todos.filter(todo => !todo.done).length,
        handleAddTodo,
        handleDeleteTodo,
        handleToggleTodo
    }
}
