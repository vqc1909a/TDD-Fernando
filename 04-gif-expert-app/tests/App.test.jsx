import { fireEvent, render, screen, waitFor, renderHook } from "@testing-library/react";
import App from "../src/App";
import { useFetchGifs } from "../src/hooks/useFetchGifs";

jest.mock('nanoid', () => ({
    nanoid: () => Math.floor(Math.random() * 100) + 1
}));

describe('tests on the component App', () => { 


  test('should show the component with the initial state  ', () => { 
    const {container} = render(<App></App>);
    expect(container).toMatchSnapshot();
  })

  test('should add the array categories when we trigger submit formulary with some value', async () => { 
    render(<App></App>);

    let category = "Naruto"
    let inputToChangeValue = screen.getByRole('textbox');
    let formToTriggerSubmit = screen.getByRole('form');

    fireEvent.change(inputToChangeValue, {target: {value: category}});
    fireEvent.submit(formToTriggerSubmit, {preventDefault(){}});

    //Simply with this renderHook and waitFor, we make the value isLoading will be true so will have gifs, or also we make the other way around, it will be the same
    const { result } = renderHook(() => useFetchGifs(category)); 

    await waitFor(() => {
      const { isLoading } = result.current;
      expect( isLoading ).toBeFalsy();
    }) 
    expect(screen.getAllByRole('generic', {name: 'card-grif'}).length).toBe(2);
  })

  test('should not add the array categories when we trigger submit formulary with empty value', async () => { 
    render(<App></App>);

    let category = ""
    let inputToChangeValue = screen.getByRole('textbox');
    let formToTriggerSubmit = screen.getByRole('form');

    fireEvent.change(inputToChangeValue, {target: {value: category}});
    fireEvent.submit(formToTriggerSubmit, {preventDefault(){}});

    const { result } = renderHook(() => useFetchGifs(category)); 

    await waitFor(() => {
      const { isLoading } = result.current;
      expect( isLoading ).toBeFalsy();
    }) 
    expect(screen.getAllByRole('generic', {name: 'card-grif'}).length).toBe(1);
  })
 })