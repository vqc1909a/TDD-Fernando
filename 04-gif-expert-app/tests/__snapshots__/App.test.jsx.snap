// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on the component App should show the component with the initial state   1`] = `
<div>
  <div>
    <h1>
      GifExpertApp
    </h1>
    <form
      aria-label="form"
    >
      <input
        placeholder="Buscar gifts"
        type="text"
        value=""
      />
    </form>
    <ol>
      <h3>
        One Puch
      </h3>
      <p>
        Loading...
      </p>
    </ol>
  </div>
</div>
`;
